File: src/application/usecase/user/update.rs
use std::sync::Arc;

use async_trait::async_trait;

use crate::application::error::AppError;
use crate::application::usecase::user::{RoleRepository, UserRepository};
use crate::domain;

pub struct UpdateUserInput {
    pub name: String,
    pub email: String,
    pub phone: String,
    pub avatar: String,
    pub role_ids: Vec<String>,
}

pub struct UpdateUserOutput {
    pub user: domain::User,
}

#[async_trait]
pub trait UpdateUserUseCase: Send + Sync {
    async fn execute(&self, id: &str, input: UpdateUserInput) -> Result<UpdateUserOutput, AppError>;
}

pub struct UpdateUserUseCaseImpl {
    user_repo: Arc<dyn UserRepository>,
    role_repo: Arc<dyn RoleRepository>,
}

impl UpdateUserUseCaseImpl {
    pub fn new(user_repo: Arc<dyn UserRepository>, role_repo: Arc<dyn RoleRepository>) -> Self {
        UpdateUserUseCaseImpl {
            user_repo,
            role_repo,
        }
    }
}

#[async_trait]
impl UpdateUserUseCase for UpdateUserUseCaseImpl {
    async fn execute(&self, id: &str, input: UpdateUserInput) -> Result<UpdateUserOutput, AppError> {
        let mut user = self.user_repo.find_by_id(id).await?;

        if input.email != user.email {
            let existing = self.user_repo.find_by_email(&input.email).await;
            if existing.is_ok() {
                return Err(AppError::from(domain::Error::AlreadyExists));
            }
        }

        super::helpers::validate_role_ids(&*self.role_repo, &input.role_ids).await?;

        user.update_info(input.name, input.email, input.phone, input.avatar, input.role_ids)?;
        self.user_repo.update(&user).await?;
        Ok(UpdateUserOutput { user })
    }
}

File: internal/application/usecase/user/update.go
package user

import (
	"context"

	"go-clean-template/internal/domain"
)

type UpdateUserInput struct {
	Name    string
	Email   string
	Phone   string
	Avatar  string
	RoleIDs []string
}

type UpdateUserOutput struct {
	User *domain.User
}

type UpdateUserUseCase interface {
	Execute(ctx context.Context, id string, input *UpdateUserInput) (*UpdateUserOutput, error)
}

type updateUserUseCase struct {
	userRepo domain.UserRepository
	roleRepo domain.RoleRepository
}

func NewUpdateUserUseCase(userRepo domain.UserRepository, roleRepo domain.RoleRepository) UpdateUserUseCase {
	return &updateUserUseCase{userRepo: userRepo, roleRepo: roleRepo}
}

func (uc *updateUserUseCase) Execute(ctx context.Context, id string, input *UpdateUserInput) (*UpdateUserOutput, error) {
	user, err := uc.userRepo.FindByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if input.Email != user.Email {
		existing, _ := uc.userRepo.FindByEmail(ctx, input.Email)
		if existing != nil {
			return nil, domain.ErrAlreadyExists
		}
	}
	if err := validateRoleIDs(ctx, uc.roleRepo, input.RoleIDs); err != nil {
		return nil, err
	}
	if err := user.UpdateInfo(input.Name, input.Email, input.Phone, input.Avatar, input.RoleIDs); err != nil {
		return nil, err
	}
	if err := uc.userRepo.Update(ctx, user); err != nil {
		return nil, err
	}
	return &UpdateUserOutput{User: user}, nil
}
